Sources/TensorFlow/Layers/Recurrent.swift

```diff
@@ -415,17 +426,146 @@ public struct RecurrentLayer<Cell: RecurrentLayerCell>: Layer {
   @differentiable(wrt: (self, inputs))
   public func lastOutput(from inputs: [Cell.TimeStepInput]) -> Cell.TimeStepOutput {
     precondition(!inputs.isEmpty, "'inputs' must be non-empty.")
-    let initialState = withoutDerivative(at: cell.zeroState(for: inputs[0]))
+    let initialState = withoutDerivative(at: cell.zeroState(for: backwardDirection ? inputs.last! : inputs.first!))
     return lastOutput(from: inputs, initialState: initialState)
   }
 }
 
+public protocol Mergeable: Differentiable, AdditiveArithmetic {
+  /// - Note: Trying to use `+` instead of `sum` results in compile error,
+  ///   when Tensor is extended for Mergeable comforting.
+  ///   Stack dump is stored at https://pastebin.com/PxZTir6F.
+  @differentiable
+  static func sum(_ lhs: Self, _ rhs: Self) -> Self
+}
+
+extension Tensor: Mergeable where Scalar: TensorFlowFloatingPoint {
+  @differentiable
+  public static func sum(_ lhs: Tensor, _ rhs: Tensor) -> Tensor {
+    lhs + rhs
+  }
+}
+
+@differentiable
+public func sum<T: Mergeable>(
+  _ first: T,
+  _ second: T
+) -> T {
+  T.sum(first, second)
+}
+
+public struct BidirectionalRecurrentLayer<Cell: RecurrentLayerCell>: Layer
+where Cell.TimeStepOutput: Mergeable {
+  public typealias Input = [Cell.TimeStepInput]
+  public typealias Output = [Cell.TimeStepOutput]
+  public typealias MergeFunction = @differentiable (Cell.TimeStepOutput, Cell.TimeStepOutput) -> Cell.TimeStepOutput
+
+  /// A wrapper around a `@differentiable` merge function.
+  ///
+  /// - Note: this exists as a workaround for runtime crashes regarding `@differentiable`function
+  ///   stored properties (TF-1122).
+  private class _MergeFunction {
+    var function: MergeFunction
+    init(_ function: @escaping MergeFunction) {
+      self.function = function
+    }
+  }
+
+  @noDerivative private let _mergeFunction: _MergeFunction
+
```

Comment:
Please add documentation comments to public stored properties.
```suggestion
  /// The forward recurrent layer.
```